### PR TITLE
add Cli Console support

### PR DESCRIPTION
**Pull request by mszczodrak**

### CODE REVIEW

File: include/cli/cli-console.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * This function pointer is called to notify about Console output.
+ *
+ * @param[in]  aBuf        A pointer to a buffer with an output.
+ * @param[in]  aBufLength  A length of the output data stored in the buffer.
+ *
+ * @returns                Number of bytes processed by the callback.
+ *
+ */
+typedef int (*otCliConsoleOutputCallback)(const char *aBuf, uint16_t aBufLength);
```
Comment (jwhui) [inline include/cli/cli-console.h]: We recently shifted to having all callbacks pass a `void *aContext` parameter that is set by the caller when setting up the callback.

File: include/cli/cli-console.h [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * Initialize the CLI CONSOLE module.
+ *
+ * @param[in]  aCallback   A callback method called to process console output.
+ *
+ */
+void otCliConsoleInit(otCliConsoleOutputCallback aCallback);
```
Comment (jwhui) [inline include/cli/cli-console.h]: Allow client to pass a `void *aContext` parameter that is passed back as a param in the callback.

File: include/cli/cli-console.h (hunk 1 above)
Comment (mszczodrak) [inline include/cli/cli-console.h]: Thanks! Fixes applied.

File: src/cli/cli_console.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <cli/cli.hpp>
+#include <cli/cli_console.hpp>
+#include <common/code_utils.hpp>
+#include <common/encoding.hpp>
```
Comment (jwhui) [inline src/cli/cli_console.cpp]: Do we need to include `encoding.hpp` here?

File: src/cli/cli_console.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+#include <common/new.hpp>
+#include <common/tasklet.hpp>
```
Comment (jwhui) [inline src/cli/cli_console.cpp]: Do we need to include `tasklet.hpp` here?

File: src/cli/cli_console.cpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+ */
+
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <cli/cli.hpp>
+#include <cli/cli_console.hpp>
+#include <common/code_utils.hpp>
```
Comment (jwhui) [inline src/cli/cli_console.cpp]: Do we need to include `code_utils.hpp` here?

File: src/cli/cli_console.hpp [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+ *   This file contains definitions for a CLI server on the CONSOLE service.
+ */
+
+#ifndef CLI_CONSOLE_HPP_
+#define CLI_CONSOLE_HPP_
+
+#include <openthread-types.h>
+#include <cli/cli_server.hpp>
+#include <cli/cli-console.h>
+#include <common/tasklet.hpp>
```
Comment (jwhui) [inline src/cli/cli_console.hpp]: Do we need to include `tasklet.hpp` here?

**Comment by jwhui** [comment]
LGTM 👍

**Pull request merged**